### PR TITLE
Use of environments in explore

### PR DESCRIPTION
**Pull request by sofiia-chorna**
Added a `environments` parameter in the `chemiscope.explore` and added an example of its usage

### CODE REVIEW

File: python/chemiscope/explore.py [hunk 1]
```diff
@@ -30,6 +30,12 @@ def explore(frames, featurize=None, properties=None, mode="default"):
         with the atomic structures. Properties can be extracted from frames with
         :py:func:`extract_properties` or manually defined by the user.
 
+    :param environments: optional. List of environments (described as
+        ``(structure id, center id, cutoff)``) to include when extracting the
+        atomic properties. Can be extracted from frames with
+        :py:func:`all_atomic_environments` (or :py:func:`librascal_atomic_environments`)
```
Comment (Luthaf) [inline python/chemiscope/explore.py:36]:
```suggestion
        :py:func:`all_atomic_environments`.
```

I think we can stop promoting `librascal_atomic_environments` now. I'll see about removing it altogether.

File: python/chemiscope/explore.py [hunk 2]
```diff
@@ -63,7 +69,7 @@ def explore(frames, featurize=None, properties=None, mode="default"):
 
 
         # Define a function for dimensionality reduction
-        def soap_kpca_featurize(frames):
+        def soap_kpca_featurize(frames, _environments):
             # Compute descriptors
```
Comment (Luthaf) [inline python/chemiscope/explore.py:73]:
```suggestion
        def soap_kpca_featurize(frames, environments):
            if environments is not None:
                raise ValueError("'environments' are not supported")
            # Compute descriptors
```

File: python/examples/6-explore.py [hunk 3]
```diff
@@ -85,6 +84,18 @@ def fetch_dataset(filename, base_url="https://zenodo.org/records/12748925/files/
...
 
+# %%
+#
+# Besides this, it is possible to run the dimentionality reduction algorithm and display
+# specific atom-centered environments. They can be manually defined by specifying a list
+# of tuples in the format ``[index of structure, index of atom, cutoff]``, as shown in
+# this example.
+#
+# Alternatively, the environments can be extracted from the frames using the function
+# :py:func:`all_atomic_environments` or :py:func:`librascal_atomic_environments`.
```
Comment (Luthaf) [inline python/examples/6-explore.py:95]:
```suggestion
# :py:func:`all_atomic_environments`.
```

File: python/examples/6-explore.py [hunk 4]
```diff
@@ -103,13 +114,14 @@ def fetch_dataset(filename, base_url="https://zenodo.org/records/12748925/files/
...
-# passed to the ``featurize`` function.
+# Define the function ``soap_kpca_featurize`` which takes two arguments
+# (``frames``, which contains the structures provided to
+# :py:func:`chemiscope.explore` and internally passed to the ``featurize`` function;
+# ``environments``,  optional aurgument with the atom-centered environments,
+# if they were provided to the :py:func:`chemiscope.explore`.
 
 
-def soap_kpca_featurize(frames):
+def soap_kpca_featurize(frames, _environments):
```
Comment (Luthaf) [inline python/examples/6-explore.py:124]:
```suggestion
def soap_kpca_featurize(frames, environments):
    if environments is not None:
        raise ValueError("'environments' are not supported")
```

File: python/examples/7-explore-advanced.py [hunk 5]
```diff
@@ -62,7 +62,7 @@ def fetch_dataset(filename, base_url="https://zenodo.org/records/12748925/files/
 # return the reduced data.
 
 
-def mace_off_tsne(frames):
+def mace_off_tsne(frames, _environments):
```
Comment (Luthaf) [inline python/examples/7-explore-advanced.py:65]:
```suggestion
def mace_off_tsne(frames, environments):
    if environments is not None:
        raise ValueError("'environments' are not supported")
```

File: python/examples/7-explore-advanced.py [hunk 6]
```diff
@@ -141,7 +141,7 @@ def mace_off_tsne(frames):
 # different mace calculator.
 
 
-def mace_mp0_tsne(frames):
+def mace_mp0_tsne(frames, _environments):
```
Comment (Luthaf) [inline python/examples/7-explore-advanced.py:144]:
```suggestion
def mace_mp0_tsne(frames, environments):
    if environments is not None:
        raise ValueError("'environments' are not supported")
```

**Review by Luthaf — COMMENTED** [review]
> Plus, maybe it makes sense to add an example of possible usage of environments inside a custom featurizer.

yes, this would be nice!

Also, I'd like to see a test of what happen if the environments are not sorted (i.e. something like `[(0, 1, 3.5), (0, 3, 3.5), (0, 2, 3.5), (1, 0, 3.5), (0, 5, 3.5)]`

File: python/chemiscope/explore.py [hunk 7]
```diff
@@ -98,13 +109,33 @@ def soap_kpca_featurize(frames):
...
+    # Pick frames and properties related to the environments if provided
+    if environments is not None:
+        # Sort environments by structure id and atom id
+        environments = sorted(environments, key=lambda x: (x[0], x[1]))
+
+        # Check structure indexes
+        unique_structures = list({env[0] for env in environments})
+        if any(index >= len(frames) for index in unique_structures):
+            raise IndexError(
+                "Some or more indices are greater than the length of the frames"
```
Comment (Luthaf) [inline python/chemiscope/explore.py:121]:
```suggestion
                "Some structure indices in 'environments' are larger than the number of frames"
```

File: python/chemiscope/explore.py [hunk 8]
```diff
@@ -158,8 +195,50 @@ def soap_pca_featurize(frames):
...
+    feats = soap.create(frames, centers=centers, n_jobs=n_jobs)
 
     # Compute pca
     pca = PCA(n_components=2)
     return pca.fit_transform(feats)
+
+
+def _extract_environment_indices(envs):
+    """
+    Extract environment indices per structure
```
Comment (Luthaf) [inline python/chemiscope/explore.py:207]:
```suggestion
    Convert from chemiscope's environements to DScribe's centers selection 
```

File: python/chemiscope/explore.py [hunk 9]
```diff
@@ -98,13 +109,33 @@ def soap_kpca_featurize(frames):
...
+        # Check structure indexes
+        unique_structures = list({env[0] for env in environments})
+        if any(index >= len(frames) for index in unique_structures):
+            raise IndexError(
+                "Some or more indices are greater than the length of the frames"
+            )
+
+        if len(unique_structures) != len(frames):
+            # Pick frames corresponding to the environments
+            frames = [frames[index] for index in unique_structures]
```
Comment (Luthaf) [inline python/chemiscope/explore.py:126]: I'm not sure this is correct: after running it, the indexes in environments will no longer match the positions of items in the `frames` list.

File: python/chemiscope/explore.py [hunk 10]
```diff
@@ -98,13 +109,33 @@ def soap_kpca_featurize(frames):
...
+
+            # Pick properties corresponding to the environments
+            properties = _extract_properties_by_environments(
+                properties, unique_structures
+            )
```
Comment (Luthaf) [inline python/chemiscope/explore.py:131]: I don't understand why this is needed. chemiscope should be fine with extra values in the structure properties, and when staying in 'structure' mode we want to be able to use all of them.

File: python/chemiscope/explore.py (hunk 9 above)
Comment (Luthaf) [reply python/chemiscope/explore.py:126]: If this is an issue with DScribe, it should be handled in the DScribe adapter.

File: python/chemiscope/explore.py (hunk 10 above)
Comment (sofiia-chorna) [reply python/chemiscope/explore.py]: Indeed, we don't need it if we provide the environemnts to `chemiscope.show`

File: python/chemiscope/explore.py (hunk 9 above)
Comment (sofiia-chorna) [reply python/chemiscope/explore.py]: Done

File: python/chemiscope/explore.py [hunk 11]
```diff
@@ -133,6 +145,29 @@ def soap_pca_featurize(frames):
...
+        # Check structure indexes
+        unique_structures = list({env[0] for env in environments})
+        if any(index >= len(frames) for index in unique_structures):
+            raise IndexError(
+                "Some structure indices in 'environments' are larger than the number of"
+                "frames"
+            )
+
+        if len(unique_structures) != len(frames):
+            # Pick frames corresponding to the environments
```
Comment (Luthaf) [inline python/chemiscope/explore.py:168]:
```suggestion
            # only include frames that are present in the user-provided environments
```

File: python/examples/7-explore-advanced.py [hunk 12]
```diff
@@ -62,7 +63,10 @@ def fetch_dataset(filename, base_url="https://zenodo.org/records/12748925/files/
 # return the reduced data.
 
 
-def mace_off_tsne(frames):
+def mace_off_tsne(frames, environments):
+    if environments is not None:
+        raise ValueError("'environments' are not supported")
```
Comment (Luthaf) [inline python/examples/7-explore-advanced.py:68]:
```suggestion
        raise ValueError("'environments' are not supported by this featurizer")
```

File: python/examples/7-explore-advanced.py [hunk 13]
```diff
@@ -178,3 +185,75 @@ def mace_mp0_tsne(frames):
...
+
+    # Get atom-centered indices from environments
+    for [env_index, atom_index, _cutoff] in environments:
+        if env_index not in grouped_envs:
+            grouped_envs[env_index] = []
+        grouped_envs[env_index].append(atom_index)
+        unique_structures.add(env_index)
+    centers = list(grouped_envs.values())
+
+    # Pick the frames corresponding to the structure indices
```
Comment (Luthaf) [inline python/examples/7-explore-advanced.py:216]:
```suggestion
    # only include frames that are present in the environments
```

**Review by Luthaf — APPROVED** [review]
Looks good, I only have a couple of suggestions on doc/comment/error messages and then this should be good to merge!

**Pull request merged**